apollo-federation/src/query_graph/graph_path.rs

```diff
@@ -1644,7 +1642,7 @@ where
                                     .get(root_kind)
                                     .copied()
                             } else {
-                                Some(last_subgraph_entering_edge_head)
+                                Some(last_subgraph_entering_edge_tail)
```

Comment:
This is an off-by-one error. `last_subgraph_entering_edge_head` corresponds to a key edge in the `GraphPath`. Key edges in the path's edge triggers are always `OpPathElement::Context`. If a context is iterated over by `check_direct_path_from_node`, that method always returns `None`, which nullifies this optimization.